File: example/helloworld/go/greeter_server.go
package main

import (
    "fmt"
    "os"
    "os/signal"
    "syscall"
    "mrpc/example/helloworld/go/helloworld"
)

func main() {
    server := helloworld.NewGreeterServer("0.0.0.0:8080")
    service := helloworld.NewGreeterService()
    server.RegisterService(service.MrpcService)
    err := server.Start()
    if err != nil {
        fmt.Println("服务器启动失败:", err)
        os.Exit(1)
    }
    fmt.Println("服务器成功启动，监听地址: 0.0.0.0:8080")

    // 信号处理
    sigChan := make(chan os.Signal, 1)
    signal.Notify(sigChan, syscall.SIGINT, syscall.SIGTERM)
    <-sigChan

    server.Stop()
    fmt.Println("服务器已停止")
}

File: example/helloworld/python/greeter_server.py
import signal
import threading
import sys
from helloworld_mrpc import GreeterServer, GreeterService, SayHelloRequest, SayHelloResponse
import mrpc  # Import mrpc to get MrpcError

# 继承GreeterService并实现SayHello方法
class GreeterServiceImpl(GreeterService):
    def SayHello(self, request: SayHelloRequest, response: SayHelloResponse) -> mrpc.MrpcError | None:
        response.message = f"Hello {request.name}"
        return None

# 信号处理函数
stop_event = threading.Event()

def signal_handler(sig, frame):
    print(f"收到信号 {sig}，准备停止服务器...")
    stop_event.set()

def main():
    server_address = "0.0.0.0:8080"
    server = GreeterServer(server_address)
    service = GreeterServiceImpl()
    server.RegisterService(service)
    err = server.Start()
    if err is not None:
        print(f"服务器启动失败: {err.message}")
        sys.exit(1)
    
    print(f"服务器成功启动，监听地址: {server_address}")
    
    # 设置信号处理
    signal.signal(signal.SIGINT, signal_handler)
    signal.signal(signal.SIGTERM, signal_handler)
    
    # 等待停止信号
    try:
        stop_event.wait()
    except KeyboardInterrupt:
        print("收到键盘中断，准备停止服务器...")
        stop_event.set()
    
    # 停止服务器
    server.Stop()
    print("服务器已停止")
    sys.exit(0)

if __name__ == "__main__":
    main()